subsys/fs/fcb/fcb.c

```diff
@@ -194,35 +205,54 @@ fcb_is_empty(struct fcb *fcb)
 /**
  * Length of an element is encoded in 1 or 2 bytes.
  * 1 byte for lengths < 128 bytes, and 2 bytes for < 16384.
+ *
+ * The storage of length has been originally designed to work with 0xff erasable
+ * flash devices and gives length 0xffff special meaning: that there is no value
+ * written; this is smart way to utilize value in non-written flash to figure
+ * out where data ends. Additionally it sets highest bit of first byte of
+ * the length to 1, to mark that there is second byte to be read.
+ * Above poses some problems when non-0xff erasable flash is used. To solve
+ * the problem all length values are xored with not of erase value for given
+ * flash:
+ *	len' = len ^ ~erase_value;
+ * To obtain original value, the logic is reversed:
+ *	len = len' ^ ~erase_value;
+ *
+ * In case of 0xff erased flash this does not modify data that is written to
+ * flash; in case of other flash devices, e.g. that erase to 0x00, it allows
+ * to correctly use the first bit of byte to figure out how many bytes are there
+ * and if there is any data at all or both bytes are equal to erase value.
  */
 int
-fcb_put_len(uint8_t *buf, uint16_t len)
+fcb_put_len(const struct fcb *fcb, uint8_t *buf, uint16_t len)
 {
 	if (len < 0x80) {
-		buf[0] = len;
+		buf[0] = len ^ ~fcb->f_erase_value;
 		return 1;
 	} else if (len < FCB_MAX_LEN) {
-		buf[0] = (len & 0x7f) | 0x80;
-		buf[1] = len >> 7;
+		buf[0] = ((len & 0x7f) | 0x80) ^ ~fcb->f_erase_value;
```

Comment:
This is a nitpick, but can `((len & 0x7f) | 0x80)` ever result in something that is not equal to just `(len | 0x80)` when you're storing it in a `uint8_t`?

Reply:
I doubt.